File: src/main/java/com/emilianomenendez/veritran/bankaccount/SavingsAccount.java
package com.emilianomenendez.veritran.bankaccount;

import com.emilianomenendez.veritran.Customer;
import com.emilianomenendez.veritran.bankaccount.transfer.BankTransfer;
import com.emilianomenendez.veritran.bankaccount.withdrawal.Withdrawal;
import com.emilianomenendez.veritran.bankaccount.withdrawal.WithdrawalLimit;
import com.emilianomenendez.veritran.money.Money;
import lombok.Getter;

@Getter
public class SavingsAccount implements BankAccount {
    private final Customer owner;
    private final String currency;
    private final WithdrawalLimit withdrawalLimit;
    private final TransactionHistory transactionHistory;
    private final BalanceTimeline balanceTimeline;

    public static SavingsAccountBuilder ownedBy(Customer owner) {
        return new SavingsAccountBuilder(owner);
    }

    public SavingsAccount(Customer owner, String currency, WithdrawalLimit withdrawalLimit, TransactionHistory transactionHistory) {
        this.owner = owner;
        this.currency = currency;
        this.withdrawalLimit = withdrawalLimit;
        this.transactionHistory = transactionHistory;
        this.balanceTimeline = new BalanceTimeline(currency, transactionHistory);
    }

    public Balance getInitialBalance() {
        return balanceTimeline.initialSnapshot();
    }

    public Balance getBalance() {
        return balanceTimeline.currentSnapshot();
    }

    public Balance getPreviousBalance() {
        return balanceTimeline.previousSnapshot();
    }

    public TransactionRecord deposit(Money amountToDeposit) {
        return Deposit.to(this)
                .amount(amountToDeposit)
                .execute();
    }

    public TransactionRecord withdraw(Money amountToWithdraw) {
        return Withdrawal.from(this)
                .limit(withdrawalLimit)
                .amount(amountToWithdraw)
                .execute();
    }

    public TransactionRecord transfer(BankAccount creditAccount, Money amountToTransfer) {
        return BankTransfer.from(this)
                .to(creditAccount)
                .amount(amountToTransfer)
                .execute();
    }

    public void addTransactionRecord(TransactionRecord transactionRecord) {
        transactionHistory.add(transactionRecord);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/domain/SavingsAccount.kt
package ar.com.flow.bankaccount.domain

import ar.com.flow.Customer
import ar.com.flow.bankaccount.domain.transaction.receipt.Receipt
import ar.com.flow.bankaccount.domain.withdrawal.CurrentFundsLimit
import ar.com.flow.bankaccount.domain.withdrawal.WithdrawalLimit
import ar.com.flow.bankaccount.ports.out.Statement
import ar.com.flow.money.Money

data class SavingsAccount(
    override val owner: Customer,
    override val currency: Currency,
    override val statement: Statement
) : BankAccount {
    var withdrawalLimit: WithdrawalLimit = CurrentFundsLimit()

    override val initialBalance: Balance
        get() = if (statement.first().isPresent) statement.first().get().amount else Balance.zero(currency)

    override val balance: Balance
        get() = statement.sum(statement.count())

    override val previousBalance: Balance
        get() = statement.sum(statement.count() - 1)

    override fun allowsWithdraw(amount: Money): Boolean {
        return withdrawalLimit.accepts(this, amount)
    }

    override fun deposit(amountToDeposit: Money) {
        Deposit.to(this)
            .amount(amountToDeposit)
            .execute()
    }

    override fun withdraw(amountToWithdraw: Money) {
        Withdrawal.from(this)
            .amount(amountToWithdraw)
            .execute()
    }

    override fun transfer(amountToTransfer: Money, creditAccount: BankAccount) {
        Transfer.from(this)
            .to(creditAccount)
            .amount(amountToTransfer)
            .execute()
    }

    override fun addReceipt(receipt: Receipt) {
        statement.add(receipt)
    }
    
    override fun equals(other: Any?): Boolean {
        if (this === other) return true

        other as SavingsAccount

        if (owner != other.owner) return false
        if (currency != other.currency) return false

        return true
    }

    override fun hashCode(): Int {
        var result = owner.hashCode()
        result = 31 * result + currency.hashCode()
        return result
    }
}